### PR TITLE
Partially revert the early feature-gatings added in #65742.

### PR DESCRIPTION
**Pull request by eddyb**
The intent here is to address #65860 ASAP (in time for beta, ideally), while leaving as much of #65742 around as possible, to make it easier to re-enable later.

Therefore, I've only kept the parts of the revert that re-add the old (i.e. non-early) feature-gating checks that were removed in #65742, and the test reverts.

I've disabled the new early feature-gating checks from #65742 entirely for now, but it would be easy to put them behind a `-Z` flag, or turn them into warnings, which would allow us to keep tests for both the early and late versions of the checks - assuming that's desirable.

cc @nikomatsakis @Mark-Simulacrum @Centril

### CODE REVIEW

**Comment by petrochenkov** [comment]
@eddyb 
Could you summarize the discussion that happened about this during the week?
Was the decision to revert until some RFC is written or something else?

I think https://github.com/rust-lang/rust/pull/65742 did the right thing, but if any specific gate causes regressions/issues in practice we can revert it or turn into a `soft_unstable` lint, which is preferable to reverting everything.

**Comment by Mark-Simulacrum** [comment]
We wanted to revert so that at least this hits nightly early in the cycle rather than in the last week for additional testing and also wanted some discussion around the general design (in particular, I feel like this is too strict -- I would like to see some exploration of allowing cfg's on stable syntax that may contain unstable syntax).

We also felt that the existing crater run was likely not hitting all the cases, specifically, since it was a nightly-enabled crater run, that meant that crates may have been detecting as such and turning features on (which would've masked errors that the PR introduced otherwise). I've been put on the hook for helping out with that crater run, and was planning on trying to do so this weekend.

**Comment by petrochenkov** [comment]
Understood.
@bors r+

**Comment by petrochenkov** [comment]
@bors p=100
Needs to be merged before  beta.
